compiler-plugin/src/main/java/io/ballerina/lib/data/yaml/compiler/YamlDataTypeValidator.java

```diff
@@ -0,0 +1,331 @@
+            case TABLE, XML -> reportDiagnosticInfo(ctx, typeSymbol.getLocation(),
+                    YamlDataDiagnosticCodes.UNSUPPORTED_TYPE);
+            case TYPE_REFERENCE -> validateExpectedType(((TypeReferenceTypeSymbol) typeSymbol).typeDescriptor(), ctx);
+            case INTERSECTION -> validateExpectedType(getRawType(typeSymbol), ctx);
+            default -> { }
+        }
+    }
+
+    private void validateTupleType(TupleTypeSymbol tupleTypeSymbol, SyntaxNodeAnalysisContext ctx) {
+        for (TypeSymbol memberType : tupleTypeSymbol.memberTypeDescriptors()) {
+            validateExpectedType(memberType, ctx);
+        }
+    }
+
+    private void validateRecordType(RecordTypeSymbol recordTypeSymbol, SyntaxNodeAnalysisContext ctx) {
+        detectDuplicateFields(recordTypeSymbol, ctx);
+
+        for (Map.Entry<String, RecordFieldSymbol> entry : recordTypeSymbol.fieldDescriptors().entrySet()) {
+            RecordFieldSymbol fieldSymbol = entry.getValue();
+            currentLocation = fieldSymbol.getLocation().orElseGet(() -> currentLocation);
+            validateExpectedType(fieldSymbol.typeDescriptor(), ctx);
+        }
+    }
+
+    private void validateUnionType(UnionTypeSymbol unionTypeSymbol,
+                                   SyntaxNodeAnalysisContext ctx) {
+        List<TypeSymbol> memberTypeSymbols = unionTypeSymbol.memberTypeDescriptors();
+        for (TypeSymbol memberTypeSymbol : memberTypeSymbols) {
+            validateExpectedType(getRawType(memberTypeSymbol), ctx);
+        }
+    }
+
+    public static TypeSymbol getRawType(TypeSymbol typeDescriptor) {
+        if (typeDescriptor.typeKind() == TypeDescKind.INTERSECTION) {
+            return getRawType(((IntersectionTypeSymbol) typeDescriptor).effectiveTypeDescriptor());
+        }
+        if (typeDescriptor.typeKind() == TypeDescKind.TYPE_REFERENCE) {
+            TypeReferenceTypeSymbol typeRef = (TypeReferenceTypeSymbol) typeDescriptor;
+            if (typeRef.typeDescriptor().typeKind() == TypeDescKind.INTERSECTION) {
+                return getRawType(((IntersectionTypeSymbol) typeRef.typeDescriptor()).effectiveTypeDescriptor());
+            }
+            TypeSymbol rawType = typeRef.typeDescriptor();
+            if (rawType.typeKind() == TypeDescKind.TYPE_REFERENCE) {
+                return getRawType(rawType);
+            }
+            return rawType;
+        }
+        return typeDescriptor;
+    }
+
+    private void reportDiagnosticInfo(SyntaxNodeAnalysisContext ctx, Optional<Location> location,
+                                      YamlDataDiagnosticCodes diagnosticsCodes) {
+        Location pos = location.orElseGet(() -> currentLocation);
+        DiagnosticInfo diagnosticInfo = new DiagnosticInfo(diagnosticsCodes.getCode(),
+                diagnosticsCodes.getMessage(), diagnosticsCodes.getSeverity());
+        if (allDiagnosticInfo.containsKey(pos) && allDiagnosticInfo.get(pos).equals(diagnosticInfo)) {
```

Comment:
What are the cases that we get the same diagnostic in the same location multiple times?

Reply:
this was added make sure there is no duplication.